zanata-war/src/main/java/org/zanata/service/impl/MergeTranslationsServiceImpl.java

```diff
@@ -0,0 +1,278 @@
+                    targetVersion.getId(), offset, batchSize,
+                    useNewerTranslation, supportedLocales,
+                    textFlowDAO, translationStateCacheImpl)
+                    .workInTransaction();
+        } catch (Exception e) {
+            log.warn("exception during copy text flow target", e);
+            return 0;
+        }
+    }
+
+    /**
+     * Check if sourceVersion or targetVersion has source document.
+     *
+     * @param sourceVersion
+     * @param targetVersion
+     */
+    private boolean isVersionsEmpty(HProjectIteration sourceVersion,
+            HProjectIteration targetVersion) {
+        if (sourceVersion.getDocuments().isEmpty()) {
+            log.error("No documents in source version {}:{}", sourceVersion
+                    .getProject().getSlug(), sourceVersion.getSlug());
+            return true;
+        }
+        if (targetVersion.getDocuments().isEmpty()) {
+            log.error("No documents in target version {}:{}", targetVersion
+                    .getProject().getSlug(), targetVersion.getSlug());
+            return true;
+        }
+        return false;
+    }
+
+    private void prepareMergeTranslationsHandle(
+            @Nonnull HProjectIteration sourceVersion,
+            @Nonnull HProjectIteration targetVersion,
+            @Nonnull MergeTranslationsTaskHandle handle) {
+        handle.setTriggeredBy(identity.getAccountUsername());
+
+        int total = getTotalProgressCount(sourceVersion, targetVersion);
+        handle.setMaxProgress(total);
+        handle.setTotalTranslations(total);
+    }
+
+    @Override
+    public int getTotalProgressCount(HProjectIteration sourceVersion,
+            HProjectIteration targetVersion) {
+        int matchCount = getTotalMatchCount(sourceVersion.getId(),
+            targetVersion.getId());
+
+        List<HLocale> locales =
+                getSupportedLocales(targetVersion.getProject().getSlug(),
+                        targetVersion.getSlug());
+
+        return matchCount * locales.size();
+    }
+
+    private int getTotalMatchCount(Long sourceVersionId, Long targetVersionId) {
+        return textFlowDAO.getSourceByMatchedContextCount(
+            sourceVersionId, targetVersionId);
+    }
+
+    public List<HLocale> getSupportedLocales(String projectSlug,
+            String versionSlug) {
+        return localeServiceImpl.getSupportedLanguageByProjectIteration(
+                projectSlug, versionSlug);
+    }
+
+    // @formatter:off
+    /**
+     * Rule of which translation should merge
+     * |          from         |       to         |   copy?   |
+     * |-----------------------|------------------|-----------|
+     * |fuzzy/untranslated     |       any        |     no    |
+     * |-----------------------|------------------|-----------|
+     * |different source text/ |                  |           |
+     * |document id            |       any        |     no    |
```

Comment:
At the moment, we merge translations that has matching `docid`, `textFlow.contentHash`, and in same language. Not sure about resId and msgctxt as this might limit the merge if its 2 different project.